Docs 2.0 content technical review (#676)
graknlabs/docs: 0001-typedb/02-dev/09-best.md

Reviewer: @flyingsilverfin : Please perform technical review.

There are only three base types (excluding roles):

* An **attribute** is a single concept with a fixed value in a given domain. That attribute is a property of the domain rather than of its 
  owners, and so is not uniquely defined by any concepts that might own it, nor does it require any owners to exist: a name of value "Bob" can be owned by either no one, one particular person, or multiple people.
* An **entity** is a single concept with an independent existence. It might practically require other entities to 
  exist, such as a car that cannot exist without its parts, but can be conceptualized without reference to those 
  other entities: a car can be imagined without considering its parts.
* A **relation** is a single concept with an existence that depends on at least one other concept. It cannot be 
  conceptualized without those concepts: it is impossible to imagine a marriage without considering its spouses.

Please see the following tips for modelling your schema:

* The best way to start modelling your database schema is to model entity types first and then proceed with connecting 
  them with relation types and assign ownership of attribute types to them later.

> * Try to create your schema as close to the data model as possible, including naming the types, building subtype 
>   hierarchy and connecting relations. That will make queries sound natural.

Author: ```suggestion
* Schema of a DB defines a vocabulary for queries to that particular DB. The closer the schema resembles the real-life data, the simpler it is to read queries for people familiar with the terminology. Consider putting extra effort into the following aspects of designing a DB schema: accurate type labels/names, an intuitive hierarchy of subtypes, and designing relations in the most meaningful way.
```

* When inserting instances of a relation, it is not necessary to define all of its **role players**. But this 
  represents an incomplete information state, such as a marriage in which only one of the spouses is known. While a 
  relation can be instantiated with only one role player, and sometimes a relation can logically exist without some of its 
  role players, it's not always the case. So an instance of a relation can imply the existence of its missing role players.
* In an n-ary relation, every combination of (**n-1**)-ary relation between its role players should be valid.
  For example, a ternary relation between an actor, a film, and a role is valid because the following binary 
  relations are valid:
  * actor and film, 
  * actor and role, 
  * film and role.
* TypeQL extends the relational functionality of SQL and strongly types it, by separating entities and relations 
  into distinct **concepts** with distinct attributes. 

    Relational tables (SQL) are homogeneous in that they can be used both to represent a list of relation's role players 
    (a join table) or a list of entity's attributes (a regular table). These two types of table are indistinguishable 

>     from the perspective of a database engine without additional context or constraints.

Author: ```suggestion
    from the perspective of an SQL database engine without additional context or constraints.
```

---
pageTitle: Best practices
keywords: api, typedb, typeql, optimization, query, design, tips
longTailKeywords: TypeDB best practice, query design, schema design
Summary: Best practices for schema and query design.
---

# Best practices

## Schema modelling

### Tips for beginners

> There are only three main base types (excluding roles):

Reviewer 2: Are you using the term `base` consistently to refer to `entity/relation/attribute` everywhere? If so, you can exclude "main"

Author: ```suggestion
There are only three base types (excluding roles):
```

> * An **attribute** is a fixed value in a given domain. That attribute is a property of the domain rather than of its 

Reviewer 2: I would include an example here if you're including one with the other two base types as well

Reviewer 2: Why not also include the sentence 'a single concept' for attributes for consistency? Does it sense to or not to?

Author: ```suggestion
* An **attribute** is a single concept with a fixed value in a given domain. That attribute is a property of the domain rather than of its 
```

>   owners, and so is not uniquely defined by any concepts that might own it, nor does it require any owners to exist.

Author: ```suggestion
  owners, and so is not uniquely defined by any concepts that might own it, nor does it require any owners to exist: a name of value "Bob" can be owned by either no one, one particular person, or multiple people.
```

* An **entity** is a single concept with an independent existence. It might practically require other entities to 
  exist, such as a car that cannot exist without its parts, but can be conceptualized without reference to those 
  other entities: a car can be imagined without considering its parts.
* A **relation** is a single concept with an existence that depends on at least one other concept. It cannot be 
  conceptualized without those concepts: it is impossible to imagine a marriage without considering its spouses.

> Please see the following tips for modelling your first schema:

Reviewer 2: I'd remove "first" - these are generally useful tips

Author: ```suggestion
Please see the following tips for modelling your schema:
```

> * Best way to start modeling your database schema is to go for entity types first and then proceed with connecting 

Reviewer 2: "The best way... is to model entity types first...and assign ownership of attribute types to them last."

Author: ```suggestion
* The best way to start modelling your database schema is to model entity types first and then proceed with connecting 
```

>   them with relation types and a assign ownership of attribute types to them later.

Author: ```suggestion
  them with relation types and assign ownership of attribute types to them later.
```

> * Try to create your schema as close to the data model as possible, including naming the types, building subtype 

Reviewer 2: This isn't entirely clear, are you trying to say all about naming? "Try to create your schema as closely to the data model as possible: use domain-specific names for types, subtypes, and connecting roles."

Author: Nope, the names (labels) is only the first of three. Hierarchy of types (subtypes) is also quite important. What kind of relations and their role players are also something to design as close to the real world in your model as possible, preferably.

Reviewer 2: I think this can be clarified somehow, but not entirely sure how! Right now it's not as powerful as it could be

Author: Please see the suggestion below - https://github.com/vaticle/docs/pull/676/files#r1168732900

  hierarchy and connecting relations. That will make queries sound natural.

> * The process is iterative. Don't forget that schema of a database can be updated. Consider testing your schema on a 

Reviewer 2: "Creating a schema is iterative"

Author: ```suggestion
* Creating a schema is an iterative process. Don't forget that schema of a database can be updated. Consider testing your schema on a 
```

  small dataset first. The sooner the schema will be adapted to all use cases and tested for edge cases — the 
  smaller amount of work will be required to adapt all other code for changes in the schema.
* Implementing drastic changes to your database schema is possible by undefining old types/rules (after undefining 

>   all subtypes and deleting all instances of data of these types) and defining new ones. You can even use Client API 

Reviewer 2: 'the Client API'

Author: ```suggestion
  all subtypes and deleting all instances of data of these types) and defining new ones. You can even use the Client API 
```

>   or TypeDB Studio to rename types without deleting them. But it might be easier just to create a new database and 

Reviewer 2: ". However, it might be easier to create a new database..."

Author: ```suggestion
  or TypeDB Studio to rename types without deleting them. However, it might be easier to create a new database and 
```

  upload your new schema and data from scratch.

> * It’s easier to debug queries using TypeDB Studio, then with any TypeDB Driver. Because Studio will handle all session 

Reviewer 2: - Typo: use "than" instead of "then"
- We don't use the term "Driver" anywhere, we should always use "client"
- Convert '.' to ','
- "..., because Studio manages sessions and transactions for you, and can visualise query results"

Author: 2. We do use it through out documentation now, especially in TypeDB Clients section. You can see the drafts of the pages on staging area. There was a selection of library vs driver and we with @shane-k-j chose the Driver. We can discuss that if you want as a separate topic.

Author: ```suggestion
* It’s easier to debug queries using TypeDB Studio, than with any TypeDB Driver, because Studio manages sessions 
```

>   and transaction management for you and even visualize the results.

Author: ```suggestion
  and transactions for you, and can visualise query results.
```

### Advanced tips

The following principles are considered being useful to take into account when you develop a more sophisticated 
database schema:

> * When inserting instances of a relation, it is not necessary to define all of its **roleplayers**. But this 

Reviewer 2: in documentation I think we can use a split word for `role players`?

  represents an incomplete information state, such as a marriage in which only one of the spouses is known. While a 
  relation can be instantiated with only one roleplayer, a relation cannot logically exist without all of its 

>   roleplayers, and so an instance of a relation always implies the existence of its missing roleplayers.

Reviewer 2: This isn't true all the time :)
You can have a `chemical-reaction` which may or may not have a `catalyst` included in the reaction. Both having it or not having it is fine.

In general I'm not sure we can make statements about completness of information of relations when role players are present or not.

Author: Fixed the wording

>   roleplayers, and so an instance of a relation always implies the existence of its missing roleplayers.

Author: ```suggestion
  represents an incomplete information state, such as a marriage in which only one of the spouses is known. While a 
  relation can be instantiated with only one role player, and sometimes a relation can logically exist without some of its 
  role players, it's not always the case. So an instance of a relation can imply the existence of its missing role players.
```

> * In an n-ary relation, every combination of (**n-1**)-ary relation between its roleplayers must be valid.

Reviewer 2: I wouldn't use "must", since these are guidelines - I'd use "should"

Author: ```suggestion
* In an n-ary relation, every combination of (**n-1**)-ary relation between its role players should be valid.
```

>   For example, a ternary relation between an actor, a film, and a role is valid because the following binary 

Reviewer 2: what ternary relation? You mean a "casting" relation?

Author: <img width="573" alt="Screenshot 2023-04-14 at 19 48 18" src="https://user-images.githubusercontent.com/48120135/232131193-bb66076b-c3fe-4232-8939-7b945f0362c7.png">

  relations are valid:
  * actor and film, 
  * actor and role, 
  * film and role.
* TypeQL extends the relational functionality of SQL and strongly types it, by separating entities and relations 

>   into distinct **concepts** with distinct attributes. 

Reviewer 2: maybe also emphaise concepts meaning "domains" as well?

Author: Sorry, I don't understand what you mean by that.

Reviewer 2: I think it's ok actually, never mind :)

    Relational tables are homogeneous in that they can be used both to represent a list of relation's roleplayers 
    (a join table) or a list of entity's attributes (a regular table). These two types of table are indistinguishable 

>     from the perspective of a database engine without additional context or constraints.

Reviewer 2: I don't understand this paragraph: tables in SQL? Or are you implying the existence of tables in TypeDB (they don't formally exist)? Indistinguishable from the SQL database engine's perspective? I think we can probably clarify the intent here and update it

Author: Relational tables... sounds pretty SQLy to me. Also the whole list element starts with "TypeQL extends the relational functionality of SQL". I can add the description of what we have in TypeDB/QL with all this strong typing. But that would be redundant IMHO.

Reviewer 2: I think if you just add "SQL" - eg. "from the perspective of an SQL database engine" it is clear? Do you think that is redundant?

Author: Fixed in https://github.com/vaticle/docs/pull/676/commits/76d2a3f51a0cef0452ad3707cb16f29d39e82eeb

* Each schema construct incurs a computational cost when traversed as part of a data query. Sorted in the order of 
  **increasing** cost, the constructs are:

1. Attributes
2. Entities

> 3. Subtypes

Reviewer 2: I'm not sure I'd actually include subtypes at all since they could exist for attributes, entities, or binary/nary relations!

Author: Haikal was the source of truth here. I think it means subtyping an entity should be more cost effective than creating a relation between two entities.

Reviewer 2: ok

4. Binary relations
5. N-ary relations
6. Rules

   Thus, it is important to consider the value of employing a rule, as the gain in query simplicity can be offset by
   the computational cost.

> * Composition (using role names to describe the functions that entities perform) is generally preferred over 

Reviewer 2: not strictly entities: entities or relations perform

Author: ```suggestion
* Composition (using role names to describe the functions that entities/relations perform) is generally preferred over 
```

  **inheritance** (using subtyping to describe those functions). This is because a type `a` should be a subtype of 
  `b` if every instance of `a` is necessarily also an instance of `b` (within the domain). For example, making `car`
  a subtype of `product` should be avoided as not every car is necessarily a product. The most general approach 
  would be to make `car` a subtype of `vehicle` and then have `car` play the role of `product` in, for instance, a 
  `purchase` relation.

  In domains where cars are always products, for instance in a car dealership database (that might also sell 
  other products: parts, add-ons, and extended warranties, etc.), it might be safe to make it a subtype of 
  `product`, however this then means that the database cannot scale to consider cars in contexts other than being 

>   products. This could be a worthwhile trade-off depending on the application, but should not be encouraged for 

Reviewer 2: I'd remove the 'inexperienced users part' and stop with stating that it could be a worthwhile tradeoff - or if you want to be more precise - "worthwhile tradeoff in very well specified models" ?

Author: Fixed

>   products. This could be a worthwhile trade-off depending on the application, but should not be encouraged for 
>   inexperienced users.

Author: ```suggestion
  products. This could be a worthwhile tradeoff in very well specified models.
```

## Developing a query

> TypeDB query processor will try to retrieve/process all required concepts in the most optimal and efficient way.

Reviewer 2: We don't really use "query processor" as a technical term, "query engine" does get used sometimes

Author: ```suggestion
TypeDB query engine will try to retrieve/process all required concepts in the most optimal and efficient way.
```

> If you have some problem with your query, we recommend trying to execute it with TypeDB Studio just to see whether 

Reviewer 2: "have a problem with"

Author: ```suggestion
If you have a problem with your query, we recommend trying to execute it with TypeDB Studio just to see whether 
```

> there is a problem with query itself, or other parts of applications logic, like connection control.

Reviewer 2: "the problem lies within the query itself..."

Author: ```suggestion
the problem lies within the query itself, or other parts of applications logic, like connection control.
```

To optimise execution time of your query try to do the following:

* Limit the number of concepts being processed by adding additional [constraints](03-match.md#patterns-overview) to 
  variables in your `match` clauses.

> * Limit the number of results returned by using pagination (limit + offset) or [aggregation](05-read.md#aggregation) 

Reviewer 2: 'aggregation' may not actually speed things up as much as limiting the number of answers does

  when possible.
* Disable inference if you don’t need it in your transaction. 

> * TypeDB process most queries in background (asynchronous). Consider sending all queries that you can before 

Reviewer 2: To keep in line with the grammar of these bullet points, you could phrase this as "Use asynchronous queries. Consider..."

> * TypeDB process most queries in background (asynchronous). Consider sending all queries that you can before 
>   starting to iterate through responses.

Author: ```suggestion
* Consider sending all queries that you can before starting to iterate through responses. TypeDB process most queries asynchronously.
```

### Async Queries

> All TypeQL queries sent to a TypeDB server, will be processed by it asynchronously in the background. Local 

Reviewer 2: remove 'in the background'

Author: ```suggestion
All TypeQL queries sent to a TypeDB server, will be processed by it asynchronously. Local 
```

> processing can take place while waiting for responses to be received. Take advantage of these asynchronous queries 

Reviewer 2: nice!

to mask network round-trip costs and increases throughput. For example, to perform 10 get queries in a transaction, 
it's best to send them all to the server before iterating over any of their answers.

Queries that return answers, such as get, insert return them as Streams/Iterators or Futures. These can then be 
awaited, or iterated, to retrieve the answers as they are computed.

<div class="note">
[Important]
When a transaction is committed or closed, all of its asynchronous queries are completed first.
</div>

### API

Data retrieved from a TypeDB database consists of concepts and delivered in the form of 
[ConceptMaps](07-response.md#conceptmap). Use the methods introduced by the TypeDB Client API to obtain more 

> information about the retrieved concept and its surroundings. Furthermore, the API allows us to traverse the 

Reviewer 2: I'd move the period a bit earlier like this:
"... to obtain more information about the retrieved concept. To explore the concept's surroundings, the Client API also has methods to traverse neighbors of a concept to obtain more insights"

Author: ```suggestion
information about the retrieved concept. To explore the concept's surroundings, the API has methods to traverse the 
```

neighbours of a specific concept instance to obtain more insights.

<div class="note">
[Important]

> When retrieving a number of concepts it is more efficient to do that with a TypeQL query.

Reviewer 2: "...a large number..."

Author: ```suggestion
When retrieving a large number of concepts it is more efficient to do that with a TypeQL query.
```

</div>

### Troubleshooting

> The following are some of the most common mistakes and nuances that we should know about and take into account to 

Reviewer 2: "The following are some of the most common mistakes and misunderstandings that could cause errors."

> The following are some of the most common mistakes and nuances that we should know about and take into account to 
> prevent potential errors or fix bugs.

Author: ```suggestion
The following are some of the most common mistakes and misunderstandings that could cause errors.
```

#### Get clause alters results

Using an optional [get](05-read.md#get-query) clause can alter the set of returned results. For example:

<!-- test-ignore -->
```typeql 
match $p isa person, has full-name $n; get $n;
```

The above query returns full-names (`$n`) owned by `$p` of the `person` type. 

Are we to expect to have a full name for every person instance in the results? No.

1. A person can have more than one attribute of type `full-name`. Every instance of attribute will get to the results.
2. A person can have no attributes of type `full-name`. In that case the person will not be represented by variable 
   `$p`. That will person will not be accounted for.
3. Finally, different people can have the same full names. In TypeDB that means different instances of `person` type 
   can own the same instance of `full-name` type. By filtering results to get only full-names you will receive a

>    deduplicated list of full-names. Because it's just a list of all attributes owned by `$p` type. 

Reviewer 2: you can merge these sentences with a comma

Author: That would belittle the accent on deduplicating the list. Full stop is better here for attention purposes.

To get complete information about all full names of every person, we need to modify the query as follows:

<!-- test-ignore -->
```typeql 
match $p isa person, has full-name $n; get $p, $n;
```

With this slight alteration (we added variable `$p` to the `get` clause) the response will consist of pairs of 
`person` type object and its owned `full-name` attribute. Because of the `person` object in the response any 

> repeated full names (represented in a database by the very same single attribute) will now be returned in pair with 

Reviewer 2: "very same single" attribute can be shortened to "very same attribute"

Author: That's an accent on attributes having only a single instance per unique value.

Reviewer 2: hmm okay, I guess that makes sense!

> their owner. If a person have two `full-name` attributes, than resulted response will contain two pairs with the 

Reviewer 2: 'If a person has'

Reviewer 2: '... attributes, then the resulting response'

Author: ```suggestion
their owner. If a person has two `full-name` attributes, then the resulting response will contain two pairs with the 
```
